### PR TITLE
Description: added documentation about PMATCH case expressions

### PR DESCRIPTION
**Pull request by thtuerk**
I finally documented PMATCH case expressions. Detailed documentation about advanced features of  `patternMatchesLib` is still missing, though.

@mn200 Could you please have a look, whether I represent your syntax correctly, please?

### CODE REVIEW

File: Manual/Description/definitions.tex [hunk 1]
```diff
@@ -787,25 +787,71 @@ \section{Quotient Types}\label{quotients}
...
+PMATCH case-expressions are simple to parse and pretty print. They
+support more features than decision-tree ones. There are guards,
+unbound variables in patterns and a wider variety of supported patterns in
+general. There is no size-blowup or surprising internal structure. Therefore,
+code generated from PMATCH case-expressions tends to be better. However,
+this comes at the price of a much heavier machinery to reason about
+PMATCH case-expressions.
+
+\subsection{Decision-tree case-expressions}
+By default, \HOL{} uses decision-tree case-expressions. Let the non-terminal $\mathit{term}$ stands for any \HOL{} term
```
Comment (mn200) [inline Manual/Description/definitions.tex]: Grammar: should be "let the non-terminal term stand for"

File: Manual/Description/definitions.tex [hunk 2]
```diff
@@ -928,6 +1007,116 @@ \section{Case Expressions}\label{CaseExp}
...
+\mathit{morecases} & ::= & \varepsilon\;\;\;|\;\;\;\texttt{|}\;\mathit{case}\;\mathit{morecases}\\
+\mathit{case} & ::= & \mathit{vardecl}\;\mathit{term}\; \mathit{guard} \;\texttt{=>}\; \mathit{term} \\
+\mathit{guard} & ::= & \varepsilon\;\;\;|\;\;\;\texttt{|}\;\texttt{where}\;\mathit{term} \\
+\mathit{vardecl} & ::= & \varepsilon\;\;\;|\;\;\;\texttt{().|}\;\;\;|\;\;\;
+\mathit{vars}\;\texttt{.|}\;\;\;|\;\;\;
+\texttt{(}\;\mathit{vars}\;\texttt{) .|} \\
+\mathit{vars} & ::= & \mathit{var}\;\;\;|\;\;\;\mathit{var}\;\texttt{,}\;\mathit{vars}
+\end{eqnarray*}
+
+If a guard is omitted, it defaults to true (\texttt{T}). Omitting the declaration of variables bound by a pattern means that all variables used in the pattern are bound. Variables, whose name start with an underscore are bound, no matter whether there appear in the list of bound variables. This is convenient for using wildcard notations.
```
Comment (mn200) [inline Manual/Description/definitions.tex]: Grammar/punctuation: should be "Variables whose names start..."

**Comment by mn200** [comment]
As *per* my last commit, the comment in the description about dtcase being "unsimplifiable" is not illustrated by the given example. Could you come up with a better example please?

**Comment by thtuerk** [comment]
I changed the example and reworded the case expression section a bit. I hope this version is better. 

I noticed, that apparently mixing literals and constructors is possible currently for decision-tree case expressions, but leads to expressions with unintended semantics. I believe this is a bug that has been introduced already some time ago. I need to check more carefully, though, before really understanding what is going on. Anyhow, I believe it is unrelated to the changes in this branch and should not affect this pull request.

**Comment by thtuerk** [comment]
The issue with mixing literal and constructor patterns I address in a pull request #378, since I believe it is unrelated.

**Comment by thtuerk** [comment]
I merged the master branch into this pull-request. This is mainly in order to get the changes of pull request #378 into this pull request. While not directly related, they have an effect on the documentation section of case expressions nevertheless.

**Comment by mn200** [comment]
Thanks for your work on this: the new example is great.

**Pull request merged**